File: packages/contracts/events.py
"""DomainEvent discriminated union contracts."""

from typing import Annotated, Any, ClassVar, Literal

from pydantic import BaseModel, ConfigDict, Field

type VersionMode = Literal["strict", "merge"]
type Actor = Literal["user", "agent", "job", "system"]
type DecisionScopeType = Literal["workspace", "draft"]


class DomainEventBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event: str
    actor: Actor = "agent"
    draft_id: str | None = None
    base_version: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None

    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ()

    @classmethod
    def reducer_version_mode(cls, scope_type: DecisionScopeType | None = None) -> VersionMode:
        return cls.version_mode


class DraftCreated(DomainEventBase):
    event: Literal["DraftCreated"] = "DraftCreated"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("draft_id",)


class DraftRenamed(DomainEventBase):
    event: Literal["DraftRenamed"] = "DraftRenamed"
    draft_id: str
    name: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("draft_id",)


class DraftCopied(DomainEventBase):
    event: Literal["DraftCopied"] = "DraftCopied"
    draft_id: str
    source_draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("draft_id",)


class DraftTrashed(DomainEventBase):
    event: Literal["DraftTrashed"] = "DraftTrashed"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("draft_id",)


class AssetImported(DomainEventBase):
    event: Literal["AssetImported"] = "AssetImported"
    asset_id: str
    job_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("asset_id", "job_id")


class AssetProbed(DomainEventBase):
    event: Literal["AssetProbed"] = "AssetProbed"
    asset_id: str
    job_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("asset_id",)


class AssetHashComputed(DomainEventBase):
    event: Literal["AssetHashComputed"] = "AssetHashComputed"
    asset_id: str
    job_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("asset_id",)


class ProxyGenerated(DomainEventBase):
    event: Literal["ProxyGenerated"] = "ProxyGenerated"
    asset_id: str
    job_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("asset_id",)


class AssetInvalidated(DomainEventBase):
    event: Literal["AssetInvalidated"] = "AssetInvalidated"
    asset_id: str
    job_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("asset_id", "job_id")


class AssetIndexReady(DomainEventBase):
    event: Literal["AssetIndexReady"] = "AssetIndexReady"
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ()


class AssetIndexFailed(DomainEventBase):
    event: Literal["AssetIndexFailed"] = "AssetIndexFailed"
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ()


class MaterialUnderstandingStarted(DomainEventBase):
    event: Literal["MaterialUnderstandingStarted"] = "MaterialUnderstandingStarted"
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ()


class MaterialUnderstandingCompleted(DomainEventBase):
    event: Literal["MaterialUnderstandingCompleted"] = "MaterialUnderstandingCompleted"
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ()


class MaterialUnderstandingFailed(DomainEventBase):
    event: Literal["MaterialUnderstandingFailed"] = "MaterialUnderstandingFailed"
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ()


class AssetLinked(DomainEventBase):
    event: Literal["AssetLinked"] = "AssetLinked"
    draft_id: str
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("draft_id", "asset_id")


class AssetUnlinked(DomainEventBase):
    event: Literal["AssetUnlinked"] = "AssetUnlinked"
    draft_id: str
    asset_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("draft_id", "asset_id")


class DecisionEventBase(DomainEventBase):
    decision_id: str
    scope_type: DecisionScopeType
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ("decision_id",)
    version_mode_by_scope: ClassVar[dict[DecisionScopeType, VersionMode]] = {
        "draft": "strict",
        "workspace": "merge",
    }

    @classmethod
    def reducer_version_mode(cls, scope_type: DecisionScopeType | None = None) -> VersionMode:
        if scope_type is None:
            return cls.version_mode
        return cls.version_mode_by_scope[scope_type]


class DecisionCreated(DecisionEventBase):
    event: Literal["DecisionCreated"] = "DecisionCreated"


class DecisionAnswered(DecisionEventBase):
    event: Literal["DecisionAnswered"] = "DecisionAnswered"


class DecisionCancelled(DecisionEventBase):
    event: Literal["DecisionCancelled"] = "DecisionCancelled"


class BriefUpdated(DomainEventBase):
    event: Literal["BriefUpdated"] = "BriefUpdated"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class ContentPlanUpdated(DomainEventBase):
    event: Literal["ContentPlanUpdated"] = "ContentPlanUpdated"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class AudioPlanUpdated(DomainEventBase):
    event: Literal["AudioPlanUpdated"] = "AudioPlanUpdated"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class CutPlanUpdated(DomainEventBase):
    event: Literal["CutPlanUpdated"] = "CutPlanUpdated"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class PostprocessPlanUpdated(DomainEventBase):
    event: Literal["PostprocessPlanUpdated"] = "PostprocessPlanUpdated"
    draft_id: str
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class TimelineVersionCreated(DomainEventBase):
    event: Literal["TimelineVersionCreated"] = "TimelineVersionCreated"
    draft_id: str
    timeline_version: int | None = None
    patch_id: str | None = None
    parent_version: int | None = None
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class TimelineVersionRestored(DomainEventBase):
    event: Literal["TimelineVersionRestored"] = "TimelineVersionRestored"
    draft_id: str
    timeline_version: int | None = None
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class TimelineValidated(DomainEventBase):
    event: Literal["TimelineValidated"] = "TimelineValidated"
    draft_id: str
    timeline_version: int | None = None
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class TimelineValidationFailed(DomainEventBase):
    event: Literal["TimelineValidationFailed"] = "TimelineValidationFailed"
    draft_id: str
    timeline_version: int | None = None
    version_mode: ClassVar[VersionMode] = "strict"
    merge_key: ClassVar[tuple[str, ...]] = ()


class PreviewRendered(DomainEventBase):
    event: Literal["PreviewRendered"] = "PreviewRendered"
    draft_id: str
    timeline_version: int
    artifact_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("timeline_version", "artifact_id")


class PreviewViewed(DomainEventBase):
    event: Literal["PreviewViewed"] = "PreviewViewed"
    draft_id: str
    preview_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("preview_id",)


class ExportCompleted(DomainEventBase):
    event: Literal["ExportCompleted"] = "ExportCompleted"
    draft_id: str
    timeline_version: int
    artifact_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("timeline_version", "artifact_id")


class MemoryCandidateExtracted(DomainEventBase):
    event: Literal["MemoryCandidateExtracted"] = "MemoryCandidateExtracted"
    candidate_id: str
    draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("candidate_id",)


class MemoryCandidateDiscarded(DomainEventBase):
    event: Literal["MemoryCandidateDiscarded"] = "MemoryCandidateDiscarded"
    candidate_id: str
    draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("candidate_id",)


class MemorySaved(DomainEventBase):
    event: Literal["MemorySaved"] = "MemorySaved"
    memory_id: str
    candidate_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("memory_id",)


class JobEnqueued(DomainEventBase):
    event: Literal["JobEnqueued"] = "JobEnqueued"
    job_id: str
    requested_by_draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("job_id",)


class JobProgress(DomainEventBase):
    event: Literal["JobProgress"] = "JobProgress"
    job_id: str
    requested_by_draft_id: str | None = None
    progress: float | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("job_id", "progress")


class JobSucceeded(DomainEventBase):
    event: Literal["JobSucceeded"] = "JobSucceeded"
    job_id: str
    requested_by_draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("job_id",)


class JobFailed(DomainEventBase):
    event: Literal["JobFailed"] = "JobFailed"
    job_id: str
    requested_by_draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("job_id",)


class JobCancelled(DomainEventBase):
    event: Literal["JobCancelled"] = "JobCancelled"
    job_id: str
    requested_by_draft_id: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("job_id",)


class PolicyRefusal(DomainEventBase):
    event: Literal["PolicyRefusal"] = "PolicyRefusal"
    refusal_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("refusal_id",)


class ProviderCallRecorded(DomainEventBase):
    event: Literal["ProviderCallRecorded"] = "ProviderCallRecorded"
    provider_call_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("provider_call_id",)


class ContextCompacted(DomainEventBase):
    event: Literal["ContextCompacted"] = "ContextCompacted"
    compaction_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("compaction_id",)


class TurnEnded(DomainEventBase):
    event: Literal["TurnEnded"] = "TurnEnded"
    turn_id: str
    draft_id: str
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("turn_id",)


class CapabilityDegraded(DomainEventBase):
    event: Literal["CapabilityDegraded"] = "CapabilityDegraded"
    degradation_id: str
    capability: str
    provider_id: str | None = None
    reason: str
    fallback: str | None = None
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("degradation_id",)


class SecurityRefusal(DomainEventBase):
    event: Literal["SecurityRefusal"] = "SecurityRefusal"
    security_refusal_id: str
    route: str
    path: str | None = None
    origin: str | None = None
    reason: str
    actor: Actor = "system"
    version_mode: ClassVar[VersionMode] = "merge"
    merge_key: ClassVar[tuple[str, ...]] = ("security_refusal_id",)


EVENT_CLASSES: tuple[type[DomainEventBase], ...] = (
    DraftCreated,
    DraftRenamed,
    DraftCopied,
    DraftTrashed,
    AssetImported,
    AssetProbed,
    AssetHashComputed,
    ProxyGenerated,
    AssetInvalidated,
    AssetIndexReady,
    AssetIndexFailed,
    MaterialUnderstandingStarted,
    MaterialUnderstandingCompleted,
    MaterialUnderstandingFailed,
    AssetLinked,
    AssetUnlinked,
    DecisionCreated,
    DecisionAnswered,
    DecisionCancelled,
    BriefUpdated,
    ContentPlanUpdated,
    AudioPlanUpdated,
    CutPlanUpdated,
    PostprocessPlanUpdated,
    TimelineVersionCreated,
    TimelineVersionRestored,
    TimelineValidated,
    TimelineValidationFailed,
    PreviewRendered,
    PreviewViewed,
    ExportCompleted,
    MemoryCandidateExtracted,
    MemoryCandidateDiscarded,
    MemorySaved,
    JobEnqueued,
    JobProgress,
    JobSucceeded,
    JobFailed,
    JobCancelled,
    PolicyRefusal,
    ProviderCallRecorded,
    ContextCompacted,
    TurnEnded,
    CapabilityDegraded,
    SecurityRefusal,
)

type EventName = Literal[
    "DraftCreated",
    "DraftRenamed",
    "DraftCopied",
    "DraftTrashed",
    "AssetImported",
    "AssetProbed",
    "AssetHashComputed",
    "ProxyGenerated",
    "AssetInvalidated",
    "AssetIndexReady",
    "AssetIndexFailed",
    "MaterialUnderstandingStarted",
    "MaterialUnderstandingCompleted",
    "MaterialUnderstandingFailed",
    "AssetLinked",
    "AssetUnlinked",
    "DecisionCreated",
    "DecisionAnswered",
    "DecisionCancelled",
    "BriefUpdated",
    "ContentPlanUpdated",
    "AudioPlanUpdated",
    "CutPlanUpdated",
    "PostprocessPlanUpdated",
    "TimelineVersionCreated",
    "TimelineVersionRestored",
    "TimelineValidated",
    "TimelineValidationFailed",
    "PreviewRendered",
    "PreviewViewed",
    "ExportCompleted",
    "MemoryCandidateExtracted",
    "MemoryCandidateDiscarded",
    "MemorySaved",
    "JobEnqueued",
    "JobProgress",
    "JobSucceeded",
    "JobFailed",
    "JobCancelled",
    "PolicyRefusal",
    "ProviderCallRecorded",
    "ContextCompacted",
    "TurnEnded",
    "CapabilityDegraded",
    "SecurityRefusal",
]

type EVENT_UNION = Annotated[
    DraftCreated
    | DraftRenamed
    | DraftCopied
    | DraftTrashed
    | AssetImported
    | AssetProbed
    | AssetHashComputed
    | ProxyGenerated
    | AssetInvalidated
    | AssetIndexReady
    | AssetIndexFailed
    | MaterialUnderstandingStarted
    | MaterialUnderstandingCompleted
    | MaterialUnderstandingFailed
    | AssetLinked
    | AssetUnlinked
    | DecisionCreated
    | DecisionAnswered
    | DecisionCancelled
    | BriefUpdated
    | ContentPlanUpdated
    | AudioPlanUpdated
    | CutPlanUpdated
    | PostprocessPlanUpdated
    | TimelineVersionCreated
    | TimelineVersionRestored
    | TimelineValidated
    | TimelineValidationFailed
    | PreviewRendered
    | PreviewViewed
    | ExportCompleted
    | MemoryCandidateExtracted
    | MemoryCandidateDiscarded
    | MemorySaved
    | JobEnqueued
    | JobProgress
    | JobSucceeded
    | JobFailed
    | JobCancelled
    | PolicyRefusal
    | ProviderCallRecorded
    | ContextCompacted
    | TurnEnded
    | CapabilityDegraded
    | SecurityRefusal,
    Field(discriminator="event"),
]
type DomainEvent = EVENT_UNION


def event_registry() -> dict[str, type[DomainEventBase]]:
    registry: dict[str, type[DomainEventBase]] = {}
    for event_class in EVENT_CLASSES:
        event_name = event_class.model_fields["event"].default
        if not isinstance(event_name, str):
            raise TypeError(f"{event_class.__name__} must declare a string event default")
        registry[event_name] = event_class
    return registry

File: go/internal/contracts/events.go
package contracts

import (
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

const TurnCancelledObservationKind = "turn_cancelled"

// TurnCancelledObservationContent 把一次取消覆盖的内存回合数写进既有 messages
// 事实源。显式计数表示新语义：只统计屏障接管的、尚未持久终态的回合；历史无计数
// 文案则由解析器按旧语义兼容，不新增表或 API/SSE 契约。
func TurnCancelledObservationContent(count int) string {
	if count < 1 {
		count = 1
	}
	return fmt.Sprintf("用户已停止当前任务（已取消 %d 个回合）。", count)
}

// ParseTurnCancelledObservation 解析计数并报告它是否采用新精确语义。历史标记和
// 异常文案保守按 1 个旧语义终态处理，避免异常数据一次核销过多用户请求。
func ParseTurnCancelledObservation(content string) (count int, exact bool) {
	const prefix = "用户已停止当前任务（已取消 "
	const suffix = " 个回合）。"
	if !strings.HasPrefix(content, prefix) || !strings.HasSuffix(content, suffix) {
		return 1, false
	}
	value := strings.TrimSuffix(strings.TrimPrefix(content, prefix), suffix)
	count, err := strconv.Atoi(value)
	if err != nil || count < 1 {
		return 1, false
	}
	return count, true
}

type VersionMode string

const (
	VersionStrict VersionMode = "strict"
	VersionMerge  VersionMode = "merge"
)

type Actor string

const (
	ActorUser  Actor = "user"
	ActorAgent Actor = "agent"
	ActorJob   Actor = "job"
)

type EventSpec struct {
	Mode               VersionMode
	WorkspaceScopeMode VersionMode
	MergeKeys          []string
	OptionalMergeKeys  []string
	DraftScope         bool
	Routes             EventRoutes
}

type EventRoutes uint8

const (
	RouteWorkspace EventRoutes = 1 << iota
	RouteDraft
)

func (routes EventRoutes) Includes(route EventRoutes) bool {
	return routes&route != 0
}

// EventRegistry 是 Go 精简核心与前端仍在使用的生命周期契约。
var EventRegistry = map[string]EventSpec{
	"DraftCreated":                   {Mode: VersionMerge, MergeKeys: []string{"draft_id"}, DraftScope: true, Routes: RouteWorkspace | RouteDraft},
	"DraftRenamed":                   {Mode: VersionMerge, MergeKeys: []string{"draft_id", "name"}, DraftScope: true, Routes: RouteWorkspace | RouteDraft},
	"DraftCopied":                    {Mode: VersionMerge, MergeKeys: []string{"draft_id"}, DraftScope: true, Routes: RouteWorkspace | RouteDraft},
	"DraftTrashed":                   {Mode: VersionMerge, MergeKeys: []string{"draft_id"}, DraftScope: true, Routes: RouteWorkspace | RouteDraft},
	"AssetImported":                  {Mode: VersionMerge, MergeKeys: []string{"asset_id", "job_id"}, Routes: RouteWorkspace | RouteDraft},
	"AssetProbed":                    {Mode: VersionMerge, MergeKeys: []string{"asset_id"}, Routes: RouteWorkspace | RouteDraft},
	"AssetLinked":                    {Mode: VersionMerge, MergeKeys: []string{"draft_id", "asset_id"}, DraftScope: true, Routes: RouteWorkspace | RouteDraft},
	"AssetUnlinked":                  {Mode: VersionMerge, MergeKeys: []string{"draft_id", "asset_id"}, DraftScope: true, Routes: RouteWorkspace | RouteDraft},
	"MaterialUnderstandingStarted":   {Mode: VersionMerge, MergeKeys: []string{"asset_id"}, OptionalMergeKeys: []string{"job_id", "attempt"}, Routes: RouteWorkspace | RouteDraft},
	"MaterialUnderstandingCompleted": {Mode: VersionMerge, MergeKeys: []string{"asset_id"}, OptionalMergeKeys: []string{"job_id", "attempt"}, Routes: RouteWorkspace | RouteDraft},
	"MaterialUnderstandingFailed":    {Mode: VersionMerge, MergeKeys: []string{"asset_id"}, OptionalMergeKeys: []string{"job_id", "attempt"}, Routes: RouteWorkspace | RouteDraft},
	"DecisionCreated":                {Mode: VersionStrict, WorkspaceScopeMode: VersionMerge, MergeKeys: []string{"decision_id"}, DraftScope: true, Routes: RouteDraft},
	"DecisionAnswered":               {Mode: VersionStrict, WorkspaceScopeMode: VersionMerge, MergeKeys: []string{"decision_id"}, DraftScope: true, Routes: RouteDraft},
	"ConversationContextCleared":     {Mode: VersionStrict, DraftScope: true, Routes: RouteDraft},
	"TimelineVersionCreated":         {Mode: VersionStrict, DraftScope: true, Routes: RouteDraft},
	"TimelineVersionRestored":        {Mode: VersionStrict, DraftScope: true, Routes: RouteDraft},
	"TimelineValidated":              {Mode: VersionStrict, DraftScope: true, Routes: RouteDraft},
	"TimelineValidationFailed":       {Mode: VersionStrict, DraftScope: true, Routes: RouteDraft},
	"PreviewRendered":                {Mode: VersionMerge, MergeKeys: []string{"timeline_version", "artifact_id"}, DraftScope: true, Routes: RouteDraft},
	"ExportCompleted":                {Mode: VersionMerge, MergeKeys: []string{"timeline_version", "artifact_id"}, DraftScope: true, Routes: RouteDraft},
	"JobEnqueued":                    {Mode: VersionMerge, MergeKeys: []string{"job_id"}, Routes: RouteWorkspace | RouteDraft},
	"JobSucceeded":                   {Mode: VersionMerge, MergeKeys: []string{"job_id"}, Routes: RouteWorkspace | RouteDraft},
	"JobFailed":                      {Mode: VersionMerge, MergeKeys: []string{"job_id"}, Routes: RouteWorkspace | RouteDraft},
	"JobCancelled":                   {Mode: VersionMerge, MergeKeys: []string{"job_id"}, Routes: RouteWorkspace | RouteDraft},
	"ProxyGenerated":                 {Mode: VersionMerge, MergeKeys: []string{"asset_id", "proxy_object_hash"}, Routes: RouteWorkspace | RouteDraft},
	"PeaksGenerated":                 {Mode: VersionMerge, MergeKeys: []string{"asset_id", "peaks_object_hash"}, Routes: RouteWorkspace | RouteDraft},
	"JobProgress":                    {Mode: VersionMerge, MergeKeys: []string{"job_id", "progress"}, OptionalMergeKeys: []string{"update_id"}, Routes: RouteWorkspace | RouteDraft},
	"PreviewViewed":                  {Mode: VersionMerge, MergeKeys: []string{"preview_id"}, DraftScope: true, Routes: RouteDraft},
}

type Event struct {
	Type        string         `json:"event"`
	Actor       Actor          `json:"actor"`
	DraftID     string         `json:"draft_id,omitempty"`
	BaseVersion *int           `json:"base_version,omitempty"`
	Payload     map[string]any `json:"payload"`
	CreatedAt   string         `json:"created_at,omitempty"`
}

func (event Event) Spec() (EventSpec, bool) {
	spec, ok := EventRegistry[event.Type]
	return spec, ok
}

// VersionMode returns the effective concurrency mode without mutating the
// registry declaration. Workspace-scoped variants opt into their mode in spec.
func (event Event) VersionMode() (VersionMode, bool) {
	spec, ok := event.Spec()
	if !ok {
		return "", false
	}
	if stringValue(event.Payload["scope_type"]) == "workspace" && spec.WorkspaceScopeMode != "" {
		return spec.WorkspaceScopeMode, true
	}
	return spec.Mode, true
}

func (event Event) Validate() error {
	spec, ok := event.Spec()
	if !ok {
		return fmt.Errorf("未知领域事件 %q", event.Type)
	}
	if !event.Actor.Valid() {
		return fmt.Errorf("事件 %s 的 actor 无效", event.Type)
	}
	if event.Payload == nil {
		return fmt.Errorf("事件 %s 缺少 payload", event.Type)
	}
	if spec.DraftScope && event.DraftID == "" && stringValue(event.Payload["scope_type"]) != "workspace" {
		return fmt.Errorf("事件 %s 缺少 draft_id", event.Type)
	}
	for _, key := range spec.MergeKeys {
		if valueForKey(event, key) == "" {
			return fmt.Errorf("事件 %s 缺少 merge key %s", event.Type, key)
		}
	}
	if event.Type == "TimelineVersionRestored" {
		mode := stringValue(event.Payload["mode"])
		if stringValue(event.Payload["checkpoint_id"]) == "" ||
			stringValue(event.Payload["restore_checkpoint_id"]) == "" {
			return errors.New("事件 TimelineVersionRestored 缺少 checkpoint_id")
		}
		if mode != "timeline" && mode != "conversation" && mode != "both" {
			return errors.New("事件 TimelineVersionRestored mode 无效")
		}
		if mode != "conversation" && stringValue(event.Payload["timeline_version"]) == "" {
			return errors.New("事件 TimelineVersionRestored 缺少 timeline_version")
		}
	}
	return nil
}

func (actor Actor) Valid() bool {
	switch actor {
	case ActorUser, ActorAgent, ActorJob:
		return true
	default:
		return false
	}
}

func (event Event) MergeKey() (string, error) {
	spec, ok := event.Spec()
	if !ok {
		return "", errors.New("事件未注册")
	}
	mode, _ := event.VersionMode()
	if mode != VersionMerge || len(spec.MergeKeys) == 0 {
		return "", nil
	}
	parts := make([]string, 0, len(spec.MergeKeys)+len(spec.OptionalMergeKeys))
	for _, key := range spec.MergeKeys {
		parts = append(parts, key+"="+valueForKey(event, key))
	}
	for _, key := range spec.OptionalMergeKeys {
		if value := valueForKey(event, key); value != "" {
			parts = append(parts, key+"="+value)
		}
	}
	return strings.Join(parts, "\x1f"), nil
}

func (event Event) JSON() ([]byte, error) {
	return json.Marshal(event)
}

func ParseEvent(data []byte) (Event, error) {
	var event Event
	if err := json.Unmarshal(data, &event); err != nil {
		return Event{}, err
	}
	return event, event.Validate()
}

func RoutesToWorkspace(event Event) bool {
	spec, ok := event.Spec()
	return ok && spec.Routes.Includes(RouteWorkspace)
}

func RoutesToDraft(event Event, draftID string) bool {
	spec, ok := event.Spec()
	if !ok || !spec.Routes.Includes(RouteDraft) {
		return false
	}
	if event.DraftID == draftID {
		return true
	}
	return stringValue(event.Payload["requested_by_draft_id"]) == draftID
}

func valueForKey(event Event, key string) string {
	if key == "draft_id" {
		return event.DraftID
	}
	return stringValue(event.Payload[key])
}

func stringValue(value any) string {
	switch typed := value.(type) {
	case string:
		return typed
	case float64:
		return fmt.Sprintf("%g", typed)
	case float32:
		return fmt.Sprintf("%g", typed)
	case int:
		return fmt.Sprintf("%d", typed)
	case int64:
		return fmt.Sprintf("%d", typed)
	case json.Number:
		return typed.String()
	default:
		return ""
	}
}

// TurnStream 领域事件名由 agent(编排)与 agentexec(领域执行)共用，置于契约层单一事实源。
const (
	TurnStreamSubagentProgress = "subagent_progress"
	TurnStreamMemoryUpdated    = "memory_updated"
)
